### PR TITLE
Sample API Design

### PR DESCRIPTION
**Pull request by Aruzal**

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    "meta": {
+        "user_id": <key>
+        "course_id": <key>
+        "upvotes" <number>
+        "num_answers": <number> // Can be used to determine next answer id
+    }
+    "title": <string>
+    "body": <string>
+}
+```
```
Comment (DarkPurple141) [inline docs/api.md]: questions should have answers as children.

File: docs/api.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+### Universities
+
+Get data for a specific university
+* /api/uni/:id
+
+Get ALL courses for a university
+* /api/uni/:uid/courses
+
+Get (n) courses for a university
+* /api/uni/:uid/courses?n=10
```
Comment (PapaPwn) [inline docs/api.md]: might be better for us to pagination on the backend, so p=1, p=2... is page 1 of course, page 2 of courses, etc.

File: docs/api.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+Get (n) questions for a course
+* /api/course/:cid/questions?n=10
+* Top (n) questions ranked by reputation?
+
+Get ALL reviews for a course
+* /api/course/:cid/reviews
+
+Get (n) questions for a course
+* /api/course/:cid/reviews?n=10
```
Comment (PapaPwn) [inline docs/api.md]: same comment as above.

File: docs/api.md [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+### Questions
+
+Get the question data for a specific question id
+* /api/course/:cid/question/:qid
+
+Get ALL answers for a question
+* /api/course/:cid/question/:qid/answers
+
+Get (n) answers for a question
+* /api/course/:cid/questions/:qid/answers?n=10
```
Comment (PapaPwn) [inline docs/api.md]: ^^^

File: docs/api.md (hunk 2 above)
Comment (DarkPurple141) [reply docs/api.md]: Agree.

File: docs/api.md [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+* Answers (Replies to Questions)
+* Reviews
+* Replies (Replies to Reviews)
+
+## Sample Schema's (json)
+
+### User
+```
+"user": {
+    "uid": <unique int>
```
Comment (DarkPurple141) [inline docs/api.md]: all ids inside objects are implicity the type of id, no need to make them specifc. Will just make frontend logic more complex.

Comment (DarkPurple141) [reply docs/api.md]: eg. User
```json
{
   "id" : "xxxx",
   "username" : "xxxx",
   "meta" : { }
}
```

File: docs/api.md [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    "profile": {
+        // Stuff for any profile data ect..
+    }
+}
+```
+
+### Universities
+```
+"university" {
+    university_id: <unqiue int>
```
Comment (DarkPurple141) [inline docs/api.md]: same as above with user.

File: docs/api.md [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    "name": <string>
+    "img_url" <string>
+}
+```
+
+### Courses
+```
+"course": {
+    "course_id": <unqiue int>
```
Comment (DarkPurple141) [inline docs/api.md]: Same as above with user

File: docs/api.md (hunk 1 above)
Comment (DarkPurple141) [reply docs/api.md]: I should add on this, from the database perspective answers should just store a single parent id, and then you do a join, but the json response should definitely at least include the ids of child answers.

File: docs/api.md [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        "review_id": <key>
+        "upvotes" <number>
+    }
+    "title": <string>
+    "body": <string>
+}
+```
+## Sample API Routes
+
+### Users
```
Comment (DarkPurple141) [inline docs/api.md]: This may have to be extended later.

File: docs/api.md [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+Get the data for a specific user
+* /api/user/:id
+
+### Universities
+
+Get data for a specific university
```
Comment (DarkPurple141) [inline docs/api.md]:
```
/api/uni should provide all universities
```

Comment (DarkPurple141) [reply docs/api.md]: (paginated)

File: docs/api.md (hunk 2 above)
Comment (DarkPurple141) [reply docs/api.md]: +1

Comment (DarkPurple141) [reply docs/api.md]: unspecified all.

File: docs/api.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        "course_id": <key>
+        "upvotes" <number>
+        "num_replies": <number> // Can be used to determine next reply id
+    }
+    "title": <string>
+    "body": <string>
+}
+```
+
+### Replies
```
Comment (DarkPurple141) [inline docs/api.md]: Wonder whether having replies as their own schema is useful. Might just add complexity. Is there anything any way a reply differs from an answer?

**Comment by DarkPurple141** [comment]
Feel free to merge

**Pull request merged**